pkg/storage/db.go

```diff
@@ -199,6 +208,79 @@ func (db *DB) ping() error {
 	return nil
 }
 
+// GeneratePartition finds the latest partition, and generate the next partition based on the previous partions time range
+func (db *DB) GeneratePartition(ctx context.Context) error {
+	stmt := "SELECT partition_end FROM partitions ORDER BY partition_begin DESC LIMIT 1"
+	row := db.sqldb.QueryRowContext(ctx, stmt)
+	var latest string
+	err := row.Scan(&latest)
+	switch err {
+	case nil:
+	case sql.ErrNoRows:
+		return db.GeneratePartitionWithTimestamp(ctx, db.now())
+	default:
+		return err
+	}
+
+	latestTime, err := time.Parse(time.RFC3339, latest)
+	if err != nil {
+		return fmt.Errorf("Invalid partition range: %s, %v", latest, err)
+	}
+
+	return db.GeneratePartitionWithTimestamp(ctx, latestTime)
+}
+
+// GeneratePartitionWithTimestamp generates the partition based on the given time
+func (db *DB) GeneratePartitionWithTimestamp(ctx context.Context, begin time.Time) error {
+	monthInterval := defaultPartitionInterval
+	begin = time.Date(begin.Year(), begin.Month(), 1, 0, 0, 0, 0, begin.Location()) // month granularity
```

Comment:
Suggestion: You can both simplify this and enable non-month precision by using https://golang.org/pkg/time/#Time.Truncate to adjust the time in conjunction with a `time.Duration` instead of an integer month value.

Reply:
We'd have to build up a duration to represent days and months, and this seemed simpler: https://stackoverflow.com/questions/36988681/time-time-round-to-day

While using the partition table will allow us to have dynamic partition intervals, we are only using the default partition interval for now. I will create a ticket to address this.